File: src/music/queue.ts
import { Message } from "discord.js";
import { nowPlaying, queues } from "./playerState";

function formatDuration(sec: number) {
    const total = Number.isFinite(sec) && sec > 0 ? sec : 0;
    const m = Math.floor(total / 60);
    const s = Math.floor(total % 60).toString().padStart(2, "0");
    return `${m}:${s}`;
}

export default async function queue(message: Message) {
    const guildId = message.guild!.id;
    const current = nowPlaying.get(guildId);
    const upcoming = queues.get(guildId) ?? [];

    if (!current && upcoming.length === 0) {
        await message.reply("Queue is empty.");
        return;
    }

    const lines: string[] = [];

    if (current) {
        lines.push(`Now Playing: **${current.title}** - ${current.artist} (${formatDuration(current.duration)})`);
    }

    if (upcoming.length > 0) {
        lines.push("");
        lines.push("Up Next:");
        for (let i = 0; i < Math.min(upcoming.length, 10); i++) {
            const track = upcoming[i];
            lines.push(`${i + 1}. **${track.title}** - ${track.artist} (${formatDuration(track.duration)})`);
        }
        if (upcoming.length > 10) {
            lines.push(`...and ${upcoming.length - 10} more`);
        }
    }

    await message.reply(lines.join("\n"));
}

File: cogs/music/queue.py
from __future__ import annotations

from discord.ext import commands

from core.cleanup import reply_and_cleanup


class QueueCog(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @commands.hybrid_command(name="queue", description="Show queue")
    async def queue(self, ctx: commands.Context) -> None:
        if not ctx.guild or not hasattr(self.bot, "music_states"):
            await reply_and_cleanup(ctx, "Queue is empty.")
            return

        state = self.bot.music_states.get(ctx.guild.id)
        if not state:
            await reply_and_cleanup(ctx, "Queue is empty.")
            return

        lines: list[str] = []
        if state.now_playing:
            lines.append(f"Now: **{state.now_playing.title}** - {state.now_playing.artist}")

        if not state.queue:
            if lines:
                lines.append("Up next: empty")
                await reply_and_cleanup(ctx, "\n".join(lines))
                return
            await reply_and_cleanup(ctx, "Queue is empty.")
            return

        lines.append("Up next:")
        for i, track in enumerate(list(state.queue)[:10], start=1):
            lines.append(f"{i}. {track.title} - {track.artist}")

        remaining = len(state.queue) - 10
        if remaining > 0:
            lines.append(f"...and {remaining} more")

        await reply_and_cleanup(ctx, "\n".join(lines))


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(QueueCog(bot))
